python/ray/includes/task.pxi

```diff
@@ -0,0 +1,208 @@
+
+    def to_string(self):
+        """Convert a Ray task specification Python object to a string.
+
+        Returns:
+            String representing the task specification.
+        """
+        return self.task_spec.get().ToFlatbuffer()
+
+    def _serialized_raylet_task(self):
+        return TaskToFlatbuffer(self.execution_dependencies.get(), self.task_spec.get())
+
+    def driver_id(self):
+        """Return the driver ID for this task."""
+        return DriverID.from_native(self.task_spec.get().DriverId())
+
+    def task_id(self):
+        """Return the task ID for this task."""
+        return TaskID.from_native(self.task_spec.get().TaskId())
+
+    def parent_task_id(self):
+        """Return the task ID of the parent task."""
+        return TaskID.from_native(self.task_spec.get().ParentTaskId())
+
+    def parent_counter(self):
+        """Return the parent counter of this task."""
+        return self.task_spec.get().ParentCounter()
+
+    def function_descriptor_list(self):
+        """Return the function descriptor for this task."""
+        cdef c_vector[c_string] function_descriptor = self.task_spec.get().FunctionDescriptor()
+        results = []
+        for i in range(function_descriptor.size()):
+            results.append(function_descriptor[i])
+        return results
+
+    def arguments(self):
+        """Return the arguments for the task."""
+        cdef:
+            CTaskSpecification *task_spec = self.task_spec.get()
+            int64_t num_args = task_spec.NumArgs()
+            int count
+        arg_list = []
+        for i in range(num_args):
+            count = task_spec.ArgIdCount(i)
+            if count > 0:
+                assert count == 1
+                arg_list.append(ObjectID.from_native(task_spec.ArgId(i, 0)))
+            else:
+                serialized_str = task_spec.ArgVal(i)[:task_spec.ArgValLength(i)]
+                obj = pickle.loads(serialized_str)
+                arg_list.append(obj)
+        return arg_list
+
+    def returns(self):
+        """Return the object IDs for the return values of the task."""
+        cdef CTaskSpecification *task_spec = self.task_spec.get()
+        return_id_list = []
+        for i in range(task_spec.NumReturns()):
+            return_id_list.append(ObjectID.from_native(task_spec.ReturnId(i)))
+        return return_id_list
+
+    def required_resources(self):
+        """Return the resource dictionary of the task."""
+        cdef:
+            unordered_map[c_string, double] resource_map = self.task_spec.get().GetRequiredResources().GetResourceMap()
+            c_string resource_name
+            double resource_value
+            unordered_map[c_string, double].iterator iterator = resource_map.begin()
+
+        required_resources = {}
+        while iterator != resource_map.end():
+            resource_name = dereference(iterator).first
+            # TODO(suquark): What is the type of the resource name (bytes, str, unicode)?
```

Comment:
In Python 2 it is required to be str (bytes). In Python 3 it is required to by str (unicode).